### PR TITLE
Android build on macOS Cross-Compile host: Fixes wrong Ninja rules generation.

### PR DESCRIPTION
**Pull request by vgorloff**
Cmake generates `rules.ninja` and `build.ninja` with macOS specific default linker options. That default options (like `-dynamiclib -Wl,-headerpad_max_install_names`) cause link errors when performing cross-compilation for Android.

Similar fixes for Window host already done. This PR provided fixes for macOS host.

### CODE REVIEW

**Comment by vgorloff** [comment]
@swift-ci please test

**Comment by vgorloff** [comment]
@compnerd Will be nice if you find a time to look on this PR.

File: cmake/modules/AddSwift.cmake [hunk 1]
```diff
@@ -1096,6 +1096,10 @@ function(_add_swift_library_single target name)
     set_target_properties("${target}"
       PROPERTIES
       INSTALL_RPATH "$ORIGIN:/usr/lib/swift/cygwin")
+  elseif("${SWIFTLIB_SINGLE_SDK}" STREQUAL "ANDROID")
+    # Cmake generates incorrect rule `$SONAME_FLAG $INSTALLNAME_DIR$SONAME` for Android build on macOS cross-compile host.
+    # Proper linker flags constructed manually. See below variable `swiftlib_link_flags_all`.
+    set_target_properties("${target}" PROPERTIES NO_SONAME true)
```
Comment (compnerd) [inline cmake/modules/AddSwift.cmake]: This should be `TRUE` rather than `true`.

File: cmake/modules/AddSwift.cmake [hunk 2]
```diff
@@ -1096,6 +1096,10 @@ function(_add_swift_library_single target name)
     set_target_properties("${target}"
       PROPERTIES
       INSTALL_RPATH "$ORIGIN:/usr/lib/swift/cygwin")
+  elseif("${SWIFTLIB_SINGLE_SDK}" STREQUAL "ANDROID")
+    # Cmake generates incorrect rule `$SONAME_FLAG $INSTALLNAME_DIR$SONAME` for Android build on macOS cross-compile host.
```
Comment (compnerd) [inline cmake/modules/AddSwift.cmake]: Nit: `CMake`.

File: cmake/modules/AddSwift.cmake [hunk 3]
```diff
@@ -1882,6 +1886,16 @@ function(add_swift_target_library name)
        NOT ${name} STREQUAL swiftRemoteMirror)
       list(APPEND swiftlib_link_flags_all "-Wl,-z,defs")
     endif()
+    # Setting back linker flags which are not supported when making Android build on macOS cross-compile host.
+    if(SWIFTLIB_SHARED)
+      if(sdk IN_LIST SWIFT_APPLE_PLATFORMS)
+        list(APPEND swiftlib_link_flags_all "-dynamiclib -Wl,-headerpad_max_install_names")
```
Comment (compnerd) [inline cmake/modules/AddSwift.cmake]: Doesn't this break if you build static?

File: cmake/modules/AddSwift.cmake [hunk 4]
```diff
@@ -1882,6 +1886,16 @@ function(add_swift_target_library name)
...
+      elseif(sdk STREQUAL ANDROID)
+        list(APPEND swiftlib_link_flags_all "-shared")
```
Comment (compnerd) [inline cmake/modules/AddSwift.cmake]: What about static builds on Android?

**Review by compnerd — CHANGES_REQUESTED** [review]

**Comment by compnerd** [comment]
CC: @Rostepher @gottesmm @gwynne @jrose-apple 

The build complexity is getting pretty absurd at this point.  I think that we need to speed up the changes to break the standard library builds into sub-configures so that we can let CMake deal with the flags for the cross-compilation.

**Comment by jrose-apple** [comment]
Again, I have no objections as long as the different target platforms continue to build in parallel.

File: cmake/modules/AddSwift.cmake (hunk 3 above)
Comment (vgorloff) [reply cmake/modules/AddSwift.cmake]: In file `cmake/modules/SwiftSharedCMakeConfig.cmake` we changed settings only for shared libraries (`CMAKE_SHARED_LIBRARY_CREATE_C_FLAGS`). This PR not altering standard CMake settings for static libs.

File: cmake/modules/AddSwift.cmake (hunk 4 above)
Comment (vgorloff) [reply cmake/modules/AddSwift.cmake]: Generated ninja rules for static libs seems OK. See default Cmake flags for macOS Host here: `/PATH_TO_HOMEBREW/Cellar/cmake/3.14.5/share/cmake/Modules/Platform/Apple-Clang.cmake`

Comment (vgorloff) [reply cmake/modules/AddSwift.cmake]: There will be another PR addressed wrong install paths for `libswiftReflection.a` and all shared libs. But that PR will not be addressed `wrong ninja rules generation` in comparison to this one.

**Comment by vgorloff** [comment]
Thank you!

Requested changes seems done.

**Review by compnerd — APPROVED** [review]
I'm not happy with this, but lets improve things after things are working.  Please squash the changes before merging.

**Comment by vgorloff** [comment]
@swift-ci please test

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by compnerd** [comment]
I believe that this is breaking the android SDK builds on Windows :-(.  We no longer pass `-soname` which means that the library soname is wrong.

**Comment by vgorloff** [comment]
Seems like another `if/else` block needed .)

**Comment by compnerd** [comment]
@vgorloff - found the issue, its a dereferencing behavior change in CMake, PR25973 addresses that.

**Pull request merged**